fw/cache.c

```diff
@@ -261,22 +261,27 @@ static DEFINE_PER_CPU(char[RESP_BUF_LEN], g_c_buf);
 static TfwStr g_crlf = { .data = S_CRLF, .len = SLEN(S_CRLF) };
 
 /* Iterate over request URI and Host header to process request key. */
-#define TFW_CACHE_REQ_KEYITER(c, uri_path, host, u_end, h_start, h_end)	\
+#define TFW_CACHE_REQ_KEYITER(c, uri_path, host, u_end, h_start,	\
+			      h_end, u_fin, h_fin)			\
+	u_fin = h_fin = false;						\
+	BUG_ON(TFW_STR_EMPTY(host));					\
 	if (TFW_STR_PLAIN(uri_path)) {					\
 		c = uri_path;						\
 		u_end = (uri_path) + 1;					\
 	} else {							\
 		c = (uri_path)->chunks;					\
 		u_end = (uri_path)->chunks + (uri_path)->nchunks;	\
 	}								\
-	if (!TFW_STR_EMPTY(host)) {					\
-		BUG_ON(TFW_STR_PLAIN(host));				\
+	if (likely(!TFW_STR_PLAIN(host))) {				\
 		h_start = (host)->chunks;				\
 		h_end = (host)->chunks + (host)->nchunks;		\
 	} else {							\
-		h_start = h_end = u_end;				\
+		h_start = host;						\
+		h_end = (host) + 1;					\
 	}								\
-	for ( ; c != h_end; ++c, c = (c == u_end) ? h_start : c)
+	for ( ; !u_fin && !h_fin;					\
+	     ++c, c = (c == u_end) ? h_start : c, u_fin = (c == u_end),	\
```

Comment:
It seems we can get rid off `u_fin` and `h_fin`, because in this version we always have `h_start` and `h_end`.

Reply:
Agree, it looks we can check only if c is not equal h_end:
`for ( ; c != h_end; ++c, c = (c == u_end) ? h_start : c)`

Reply:
No we can't do it because sometimes `c = uri_path;` == h_end. For `GET http://user@-x/ HTTP/1.1\r\nHost: bad.com\r\n\r\n` for example

Reply:
Now I understand the problem: `uri_path` is placed immediately after `host` field in `TfwHttpReq`, and after `h_end = (host) + 1;` h_end points to uri_path. Therefore `c == uri_path == h_end` and the loop is not executed.

Reply:
Maybe in such a way we can get rid of u_fin and h_fin?
`for ( ; c != u_end || ((u_end == h_end) ? 0 : (c = h_start, u_end = h_end, 1)); ++c)`